### PR TITLE
🎉 New destination: Rockset

### PR DESCRIPTION
**Pull request by samecrowder**
## What
Add a new destination supporting [Rockset](www.rockset.com), the industry-leading real-time database in the cloud!

## How
Implemented the destination using the `generate.sh` script and according to standard patterns.

## Pre-merge Checklist
- [X] Issue acceptance criteria met
- [X] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [X] Secrets are annotated with `airbyte_secret` in the connector's spec
- [ ] Credentials added to Github CI if needed and not already present. [instructions for injecting secrets into CI](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#using-credentials-in-ci). 
- [X] Unit & integration tests added as appropriate (and are passing)
    * Community members: please provide proof of this succeeding locally e.g: screenshot or copy-paste acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [X] `/test connector=connectors/<name>` command as documented [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-an-existing-connector) is passing. 
    * Community members can skip this, Airbyters will run this for you. 
- [X] Code reviews completed
- [X] Documentation updated 
    - [X] `README.md`
    - [X] `docs/SUMMARY.md` if it's a new connector
    - [X] Created or updated reference docs in `docs/integrations/<source or destination>/<name>`.
    - [X] Changelog in the appropriate page in `docs/integrations/...`. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [X] `docs/integrations/README.md` contains a reference to the new connector
    - [X] Build status added to [build page](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/builds.md)
- [X] Build is successful
- [X] Connector version bumped like described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/contributing-to-airbyte/building-new-connector#updating-a-connector)
- [X] No major blockers
- [ ] PR merged into master branch
- [ ] Follow up tickets have been created
- [ ] Associated tickets have been closed & stakeholders notified

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks for the contribution @samecrowder!

**Comment by samecrowder** [comment]
> Thanks for the contribution @samecrowder!

HI @marcosmarxm , what's the next step in the process for getting this merged? I know some of the checklist is yet incomplete, but it seems that some of it falls on the Airbyte side of things. Thanks for the prompt reply!

**Comment by sherifnada** [comment]
@samecrowder I'll leave a review today with next steps! The process should roughly be: 

1. you & i iterate on the connector code on this PR
2. we (airbyte) create a sandbox account on rockset which allows us to continuously run integration tests in CI
3. merge 
4. sit back and receive OSS fame and glory

File: airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetDestination.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+  public static void main(String[] args) throws Exception {
+    new IntegrationRunner(new RocksetDestination()).run(args);
+  }
+
+  @Override
+  public AirbyteConnectionStatus check(JsonNode config) {
+    try {
+      String workspace = config.get(WORKSPACE_ID).asText();
+      RocksetUtils.createWorkspaceIfNotExists(RocksetUtils.clientFromConfig(config), workspace);
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetDestination.java:67]: could we also create a temporary table, insert a record, then delete it to make sure we have write permissions?

File: airbyte-integrations/connectors/destination-rockset/src/test-integration/java/io/airbyte/integrations/destination/rockset/RocksetDestinationAcceptanceTest.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  protected JsonNode getConfig() throws IOException {
+    return Jsons.deserialize(MoreResources.readResource("secrets/config.json"));
+  }
+
+  @Override
+  protected JsonNode getFailCheckConfig() throws Exception {
+    return Jsons.jsonNode(
+        ImmutableMap.builder()
+            .put("workspace", "commons")
+            .put("api_key", "nope nope nope")
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-rockset/src/test-integration/java/io/airbyte/integrations/destination/rockset/RocksetDestinationAcceptanceTest.java:90]: my favorite part of writing integration tests is phrasing these messages

File: airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetWriteApiConsumer.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    this.catalog = catalog;
+    this.outputRecordCollector = outputRecordCollector;
+  }
+
+  @Override
+  public void start() throws Exception {
+    this.client = new RocksetClient(apiKey, APISERVER_URL);
+
+    RocksetUtils.createWorkspaceIfNotExists(client, workspace);
+        catalog.getStreams()
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetWriteApiConsumer.java:50]: indentation

File: airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetWriteApiConsumer.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                    .collect(Collectors.toList());
+
+    for (String cname : collectionNames) {
+      RocksetUtils.createCollectionIfNotExists(client, workspace, cname);
+      RocksetUtils.waitUntilCollectionReady(client, workspace, cname);
+    }
+  }
+
+  @Override
+  public void accept(AirbyteMessage message) throws Exception {
```
Comment (sherifnada) [inline airbyte-integrations/connectors/destination-rockset/src/main/java/io/airbyte/integrations/destination/rockset/RocksetWriteApiConsumer.java:73]: wow, really simple!

**Review by sherifnada — COMMENTED** [review]
@samecrowder LGTM with one addition requested in the `check` method!

**Comment by sherifnada** [comment]
@samecrowder acceptance tests are failing for me. Can you try running `./gradlew :airbyte-integrations:connectors:destination-rockset:integrationTest` from the repo root to reproduce the failures on your side and iterate to make it succeed? 

Issues I'm seeing: 

1. The destination does not transform stream names with illegal characters into a charset accepted by rockset
2. the connection check is failing
3. linebreak characters are not being handled properly (not sure exactly what the problem is but the test is failing)

**Comment by samecrowder** [comment]
> @samecrowder acceptance tests are failing for me. Can you try running `./gradlew :airbyte-integrations:connectors:destination-rockset:integrationTest` from the repo root to reproduce the failures on your side and iterate to make it succeed?
> 
> Issues I'm seeing:
> 
> 1. The destination does not transform stream names with illegal characters into a charset accepted by rockset
> 2. the connection check is failing
> 3. linebreak characters are not being handled properly (not sure exactly what the problem is but the test is failing)

thank you yes, working on those now

**Comment by samecrowder** [comment]
@sherifnada question for you: The one integration test that I'm still failing locally is `testIncrementalSync`. Looking into the record comparison, I see that both expected and actual have 6 records and that (not accounting for ordering) the records all match.

Why exactly does DestinationAcceptanceTest.java require the record orders to be the same after retrieving from the destination? Is that an airbyte correctness spec more broadly, or just a requirement of the test suite?

**Comment by sherifnada** [comment]
Ordering is not required for correctness but I believe implementation uses a List internally for comparing expected and actual records (instead of a set to allow for detecting repeated records)

**Comment by sherifnada** [comment]
If your PR is up to date I can fix the test implementation

**Comment by sbaldwin-rs** [comment]
> Ordering is not required for correctness but I believe implementation uses a List internally for comparing expected and actual records (instead of a set to allow for detecting repeated records)

Perhaps the implementation could move to a multiset structure to avoid this issue.

**Comment by sherifnada** [comment]
@sbaldwin-rs yup, exactly what I had in mind!

**Comment by samecrowder** [comment]
@sherifnada Alright I just pushed everything so it's up to date. thank you! ping me if anything else doesn't look right

**Comment by sherifnada** [comment]
@samecrowder removing ordering from tests has proved harder than I thought. It's not impossible but I'm not able to spend time on it at the moment. Could we use the [`ORDER BY` directive](https://rockset.com/docs/commands/#order-by-clause) in SQL to get around the ordering issue? 

If so, if you can make the change and upload a screenshot I'd be happy to merge this in. Apologies for the late response again!

**Comment by marcosmarxm** [comment]
@samecrowder thanks for the contribution! can you sign the CLA and provide a screenshot demonstrating tests are passing locally?

**Comment by sbaldwin-rs** [comment]
Hi, I am hoping to get this unblocked so it can be merged. Some of the links outlined in the pre-merge checklist are stale. Would it be possible to update the pre-merge checklist with the newest tasks and updated links?

**Comment by marcosmarxm** [comment]
I'm closing this in favor of https://github.com/airbytehq/airbyte/issues/8556

**Pull request closed without merging**